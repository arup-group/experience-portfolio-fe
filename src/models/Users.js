import { types, flow } from "mobx-state-tree";
import * as api from "../utils/api";
import { Projects } from "./Projects";

export const IndividualUser = types
  .model("CurrentUser", {
    DisciplineName: types.string,
    Email: types.string,
    GradeLevel: types.number,
    JobTitle: types.string,
    LocationName: types.string,
    StaffID: types.number,
    StaffName: types.string,
    StartDate: types.string,
    careerStart: types.maybeNull(types.string),
    committees: types.maybeNull(types.string),
    highLevelDescription: types.maybeNull(types.string),
    imgUrl: types.maybeNull(types.string),
    nationality: types.maybeNull(types.string),
    professionalAssociations: types.maybeNull(types.string),
    publications: types.maybeNull(types.string),
    qualifications: types.maybeNull(types.string),
    valueStatement: types.maybeNull(types.string),
  })
  .actions((self) => ({
    changeStaffName(newName) {
      self.StaffName = newName;
    },
  }));

export const User = types
  .model("User", {
    projList: types.optional(Projects, {}),
    currentUser: types.array(IndividualUser),
    isLoading: true,
  })
  .views((self) => ({
    get showUser() {
      return self.currentUser;
    },
  }))
  .actions((self) => ({
    addUser(newUser) {
      self.currentUser = [];
      self.currentUser.push(newUser);
    },
    removeUser() {
      self.currentUser = [];
    },
    editUserMetaData: flow(function* editUserMetaData(staffID, newMetaData) {
      try {
        const data = yield api.patchUserMetaData(staffID, newMetaData);
        for (const [key, value] of Object.entries(data)) {
          self.currentUser[0][key] = value;
        }
      } catch (error) {
        console.log("something went wrong on the patch", error);
      }
    }),
    fetchMetaData: flow(function* fetchMetaData(staffID) {
      try {
        self.currentUser = [];
        const data = yield api.getUsers(staffID);
        self.currentUser.push(data);
        self.projList.fetchProjects(staffID);
        self.isLoading = false;
      } catch (error) {
        console.log("something went wrong on the fetch", error);
      }
    }),
  }));
