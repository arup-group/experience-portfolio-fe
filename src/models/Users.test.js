import { getSnapshot, onSnapshot } from "mobx-state-tree";
import { IndividualUser, User } from "./Users";

it("should create an instance of the model", () => {
  const user = IndividualUser.create({
    DisciplineName: "Structural Engineering",
    Email: "Sam.Styles@arup.com",
    GradeLevel: 6,
    JobTitle: "Senior Engineer",
    LocationName: "Manchester Office",
    StaffID: 37704,
    StaffName: "Samuel Styles",
    StartDate: "2007-09-05T23:00:00.000Z",
    careerStart: null,
    committees: null,
    highLevelDescription: null,
    imgURL: null,
    nationality: null,
    professionalAssociations: null,
    publications: null,
    qualifications: null,
    // valueStatement: null,
  });
  expect(user.GradeLevel).toBe(6);
  expect(user.committees).toBe(null);
  expect(user.valueStatement).toBe(null);
  user.changeStaffName("Sam Styles");
  expect(user.StaffName).toBe("Sam Styles");
});
it("should create an individual user for the session", () => {
  const indUser = User.create();
  const states = [];
  onSnapshot(indUser, (snapshot) => {
    states.push(snapshot);
  });
  indUser.addUser({
    DisciplineName: "Structural Engineering",
    Email: "Sam.Styles@arup.com",
    GradeLevel: 6,
    JobTitle: "Senior Engineer",
    LocationName: "Manchester Office",
    StaffID: 37704,
    StaffName: "Samuel Styles",
    StartDate: "2007-09-05T23:00:00.000Z",
    careerStart: null,
    committees: null,
    highLevelDescription: null,
    imgURL: null,
    nationality: null,
    professionalAssociations: null,
    publications: null,
    qualifications: null,
    valueStatement: null,
  });

  expect(indUser.currentUser.length).toBe(1);
  expect(indUser.currentUser[0].nationality).toBe(null);
  indUser.currentUser[0].changeStaffName("Sam Styles");
  expect(indUser.currentUser[0].StaffName).toBe("Sam Styles");
  expect(getSnapshot(indUser)).toMatchSnapshot();

  expect(states).toMatchSnapshot();
});
