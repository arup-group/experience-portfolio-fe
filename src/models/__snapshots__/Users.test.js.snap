// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create an individual user for the session 1`] = `
Object {
  "currentUser": Array [
    Object {
      "DisciplineName": "Structural Engineering",
      "Email": "Sam.Styles@arup.com",
      "GradeLevel": 6,
      "JobTitle": "Senior Engineer",
      "LocationName": "Manchester Office",
      "StaffID": 37704,
      "StaffName": "Sam Styles",
      "StartDate": "2007-09-05T23:00:00.000Z",
      "careerStart": null,
      "committees": null,
      "highLevelDescription": null,
      "imgURL": null,
      "nationality": null,
      "professionalAssociations": null,
      "publications": null,
      "qualifications": null,
      "valueStatement": null,
    },
  ],
  "isLoading": true,
}
`;

exports[`should create an individual user for the session 2`] = `
Array [
  Object {
    "currentUser": Array [
      Object {
        "DisciplineName": "Structural Engineering",
        "Email": "Sam.Styles@arup.com",
        "GradeLevel": 6,
        "JobTitle": "Senior Engineer",
        "LocationName": "Manchester Office",
        "StaffID": 37704,
        "StaffName": "Samuel Styles",
        "StartDate": "2007-09-05T23:00:00.000Z",
        "careerStart": null,
        "committees": null,
        "highLevelDescription": null,
        "imgURL": null,
        "nationality": null,
        "professionalAssociations": null,
        "publications": null,
        "qualifications": null,
        "valueStatement": null,
      },
    ],
    "isLoading": true,
  },
  Object {
    "currentUser": Array [
      Object {
        "DisciplineName": "Structural Engineering",
        "Email": "Sam.Styles@arup.com",
        "GradeLevel": 6,
        "JobTitle": "Senior Engineer",
        "LocationName": "Manchester Office",
        "StaffID": 37704,
        "StaffName": "Sam Styles",
        "StartDate": "2007-09-05T23:00:00.000Z",
        "careerStart": null,
        "committees": null,
        "highLevelDescription": null,
        "imgURL": null,
        "nationality": null,
        "professionalAssociations": null,
        "publications": null,
        "qualifications": null,
        "valueStatement": null,
      },
    ],
    "isLoading": true,
  },
]
`;
